warehouse/query-core/src/test/java/datawave/query/jexl/visitors/IngestTypePruningVisitorTest.java

```diff
@@ -0,0 +1,419 @@
+        }
+    }
+
+    @Test
+    public void testIntersectionsWithMarkers() {
+        // all marker node types
+        //  @formatter:off
+        String[] queries = {
+                        "A == '1' && ((_Bounded_ = true) && (B >= '0' && B <= '10'))",
+                        "A == '1' && ((_Delayed_ = true) && (B == '2'))",
+                        "A == '1' && ((_Delayed_ = true) && (A == '1' || B == '2'))",
+                        "A == '1' && ((_Delayed_ = true) && (A == '1' && B == '2'))",
+                        "A == '1' && ((_Eval_ = true) && (B == '2'))",
+                        "A == '1' && ((_List_ = true) && ((id = 'some-bogus-id') && (field = 'B') && (params = '{\"values\":[\"a\",\"b\",\"c\"]}')))",
+                        "A == '1' && ((_Term_ = true) && (B == '2'))",
+                        "A == '1' && ((_Value_ = true) && (B =~ 'ba.*'))",
+                        "A == '1' && ((_Value_ = true) && (A =~ 'ab.*' || B =~ 'ba.*'))",
+                        "A == '1' && ((_Value_ = true) && (A =~ 'ab.*' && B =~ 'ba.*'))"
+        };
+        //  @formatter:on
+
+        for (String query : queries) {
+            test(query, query);
+        }
+
+        // same queries as above, test pruning
+        //  @formatter:off
+        queries = new String[] {
+                        "A == '1' && ((_Bounded_ = true) && (C >= '0' && C <= '10'))",
+                        "A == '1' && ((_Delayed_ = true) && (C == '2'))",
+                        "A == '1' && ((_Eval_ = true) && (C == '2'))",
+                        "A == '1' && ((_List_ = true) && ((id = 'some-bogus-id') && (field = 'C') && (params = '{\"values\":[\"a\",\"b\",\"c\"]}')))",
+                        "A == '1' && ((_Term_ = true) && (C == '2'))",
+                        "A == '1' && ((_Value_ = true) && (C =~ 'ba.*'))"
+        };
+        //  @formatter:on
+
+        for (String query : queries) {
+            test(query, null);
+        }
+    }
+
+    @Test
+    public void testMultiFieldedMarkers() {
+        // case 1: delayed intersection of non-intersecting ingestTypes should remove itself
+        String query = "((_Delayed_ = true) && (A == '1' && C == '2'))";
+        test(query, null);
+
+        // case 2: overlapping ingestTypes
+        query = "A == '1' && ((_Delayed_ = true) && (B == '1' || C == '2'))";
+        test(query, "A == '1' && ((_Delayed_ = true) && (B == '1'))");
+
+        // case 3: non-intersecting ingestTypes (function removes itself)
+        query = "A == '1' && ((_Delayed_ = true) && (A == '1' && C == '2'))";
+        test(query, null);
+
+        // case 4: unknown field and how that works
+        query = "((_Delayed_ = true) && (A == '1' && D == '2'))";
+        test(query, null);
+    }
+
+    @Test
+    public void testDelayedBoundedMarker() {
+        String query = "((_Delayed_ = true) && ((_Bounded_ = true) && (A > '2' && A < '4')))";
+        test(query, query);
+
+        // C term drives pruning of double nested marker
```

Comment:
I've been reading through this trying to figure out how to set the type we prune by, but this is suggesting we just pick the types for the first term we see?   If so, it would be good to indicate this in the description of the parameter, e.g., "should the query prune fields by the ingest type of the first term we happen to see in the query".

Reply:
Got it, thanks